fixes #10926 - one click boot off CD for vmware

timogoebel opened the pull request:
This PR adds the possibility to boot off CD rather than network/image via compute resources (vmware). It's still work in progress. I would love some early feedback on this.

I also need to add tests.
Of course, this needs fog changes: https://github.com/fog/fog-vsphere/pull/13

Waiting for a new fog release.

What do you think?

domcleal commented:
Yep, I like it.  I looked at it doing this for libvirt once, you can see my efforts at https://github.com/domcleal/foreman_bootdisk/commits/3287-iso-libvirt and they look very similar.

We should get support for adding new provisioning methods via the plugin API into Foreman core which will make this more flexible and work properly in development environments (editing constants is messy), plus anything we can get into Fog or Foreman core models that's appropriate will help.

fog-libvirt gained some support for uploading ISOs in https://github.com/fog/fog-libvirt/commit/b043ae95ad0a1ebcb3ebf8eaebf54afeb0fb353c so you may want to look at that as a model too, to keep the methods similar.

angystardust commented:
Great! These would also solve http://projects.theforeman.org/issues/3287

timogoebel commented:
@domcleal : Thanks for taking a look at it, glad you like it. :-) I did not know your efforts, I liked some ideas and have boldly borrowed them.
I think, I have to check out the plugin api - should not be that hard to create a patch that adds support for adding new provisioning methods. I am not sure however, how to fix the JS that hides / shows the provisioning method specific form elements.

bechtoldt commented:
@timogoebel cool stuff!

timogoebel commented:
Created a PR to add the new provisioning method via the plugin api: https://github.com/theforeman/foreman/pull/2945

angystardust commented:
Any (good) news about this one? This feature will greatly help us migrating from our current cobbler setup :fireworks:

timogoebel commented:
@angystardust: Thanks for your interest. All the requirements for this are merged into foreman's develop branch and the code should work as of now. I need to test this more thoroughly, fix some minor (mostly style) issues in the code and add tests, before this is ready to go.

timogoebel commented:
I had some time to work on this during the weekend. The code works well now in our production environment, I fixed some minor glitches. In addition I added some tests.

There are still some things to do:

- [ ] I identified an issue in fog-vsphere with iso-uploading: https://github.com/fog/fog-vsphere/pull/23
- [ ] It is not possible to rebuild a bootdisk provisioned host as of yet.

angystardust commented:
Great work, @timogoebel!!!
So do we need to wait for another fog-vsphere release to push this one in?
I would be very happy to test this but I'm not so confident in using the nightly branch...maybe it's better to wait for 1.11 RC1 (hoping it will be merged before the branching... @domcleal )

timogoebel commented:
You could ask on the fog-vsphere PR for a release. I have done that too often lately and feel bad to ask yet another time.
I have rpms which add a lot of vsphere features on top of 1.10.1. I could make them available if you want.
I am still not happy with the rebuild code. It currently runs when you rewrite the config via the proxies. Id prefer that to happen when you re enable the host for building. @domcleal: Any ideas?

> On 25.01.2016, at 19:33, angystardust <notifications@github.com> wrote:
> 
> Great work, @timogoebel!!!
> So do we need to wait for another fog-vsphere release to push this one in?
> I would be very happy to test this but I'm not so confident in using the nightly branch...maybe it's better to wait for 1.11 RC1 (hoping it will be merged before the branching... @domcleal )
> 
> —
> Reply to this email directly or view it on GitHub.
>

domcleal commented:
> I am still not happy with the rebuild code. It currently runs when you rewrite the config via the proxies. Id prefer that to happen when you re enable the host for building. @domcleal: Any ideas?

You'd want to add the code to Host::Managed when the mode goes from build=false to build=true.  The easiest way is the after_build hook (added near the top of the Host::Managed source file), so in ForemanBootdisk::HostExt add:

    after_build :rebuild_with_bootdisk

and then define that method to add the disk to the host.  You can also add a `before_provision` to remove it again later (that'll be called when it leaves build mode, right at the end of the provision).

timogoebel commented:
@domcleal : Thanks a lot for the hint. :-) I think, the before_provision will not work because the host has not rebooted into the os when the event is triggered. I think, the os installer should just eject the cd itself.

Besides the fog-vsphere update, I think, my work is done here for now. Looking forward for some feedback.

angystardust commented:
@timogoebel, a newer release of fog-vsphere (0.6.0) was just released! :smile: 
Would be great to have a repo to test all the vmware feature you're adding in foreman...

timogoebel commented:
@angystardust: https://github.com/theforeman/foreman/pull/3135
Let me see, what I can do about the repo next week. :-)

domcleal commented on README.md:
```diff
@@ -47,8 +47,10 @@ gPXE images are unsupported due to lack of initrd support.
-the image.  Hosts will be identified by their MAC or IP address to provide
+the image by hand.  Hosts will be identified by their MAC or IP address to provide
 the correct provisioning template if the host is in build mode.
+On VMWare compute resources it is possible to create a new host with a cd-drive
```
Can you expand on this?  Split it into a new paragraph and state that you use it by selecting the Bootdisk Based provisioning method under the OS tab when creating the host, or by setting `provision_method: "bootdisk"` using the API.

domcleal commented on app/models/concerns/foreman_bootdisk/compute_resources/vmware.rb:
```diff
@@ -0,0 +1,61 @@
+          'local_path' => iso,
+          'datacenter' => dc.name,
+          'datastore' => bootdisk_datastore(vm_uuid),
+          'upload_directory' => 'foreman_isos'
```
I'm unfamiliar with this, but do we have to worry about creating the directory?

domcleal commented on app/models/concerns/foreman_bootdisk/orchestration/compute.rb:
```diff
@@ -0,0 +1,82 @@
+      def queue_bootdisk_compute
+        return unless compute? && errors.empty? && new_record?
+        return unless provision_method == 'bootdisk'
+        queue.create(name: _('Generating iso image for %s') % self, priority: 5,
```
Nitpick, `ISO` here and twice below.  Also `CDROM` below.

domcleal commented on app/models/concerns/foreman_bootdisk/orchestration/compute.rb:
```diff
@@ -0,0 +1,82 @@
+        logger.info 'Generating iso image for %s' % name
+        bootdisk_generate_iso_image
+      rescue => e
+        failure _("Failed to generate iso image for instance %{name}: %{message}\n ") % { name: name, message: e.message }, e
```
These failure strings (here and below) probably shouldn't have trailing whitespace as it causes translators problems, remove the `\n `.

domcleal commented on app/models/concerns/foreman_bootdisk/orchestration/compute.rb:
```diff
@@ -0,0 +1,82 @@
+      end
+
+      def bootdisk_isofile
+        File.join(Dir.tmpdir, "#{name}.iso")
```
This could potentially be attacked as tmpdir may be shared and you're `mv`ing over the top with a predictable filename.

It might be best to change this to create a temporary directory ([Dir.mktmpdir](http://ruby-doc.org/stdlib-2.0.0/libdoc/tmpdir/rdoc/Dir.html)), store the path in an instance variable (i.e. `@bootdisk_isodir ||= Dir.mktmpdir`) and copy the ISO into that instead.

domcleal commented on lib/foreman_bootdisk/engine.rb:
```diff
@@ -28,9 +28,25 @@ class Engine < ::Rails::Engine
+    # content twice.
+    assets_to_precompile =
+      Dir.chdir(root) do
+      Dir['app/assets/javascripts/**/*', 'app/assets/stylesheets/**/*'].map do |f|
```
Indentation's a bit off here, should be indented by two more chars.

domcleal commented on lib/foreman_bootdisk/engine.rb:
```diff
@@ -41,6 +57,7 @@ class Engine < ::Rails::Engine
 
         allowed_template_helpers :bootdisk_chain_url, :bootdisk_raise
         apipie_documented_controllers ["#{ForemanBootdisk::Engine.root}/app/controllers/foreman_bootdisk/api/v2/*.rb"]
+        provision_method 'bootdisk', 'Bootdisk Based'
```
Add `N_(..)` here, `N_('Bootdisk Based')` to ensure we translate it.

domcleal commented:
Very impressive!  Looks good.  Let me know if you're unable to get a repo together and I can help, as I'll need @angystardust or somebody else's help to test this.

timogoebel commented:
I updated the code a bit. There is one minor issue remaining:
Even though I have "reboot --eject" in my kickstart provision template, anaconda does not eject the cd when rebooting the vm. So the VM tries to boot another time from cd, ipxe fails (because the token was invalidated) and the vm finally boots into the os on the harddisk. This is acceptable behaviour though, I think. Does anyone have an idea how to fix this?

@angystardust: Are you still willing to test this? Would the nightly packages be something you'd be comfortable with? If not, I have uploaded some patched 1.10.1 rpm's here: https://tgoebel.de/trunk/tfm-packages/ . Please note, that they will probably break upgrading to any "official" packages. So remove these packages and install "clean" ones before any upgrade.

To actually test this, you'd want to do something like this:
```
wget -O /tmp/bootdisk.patch https://patch-diff.githubusercontent.com/raw/theforeman/foreman_bootdisk/pull/19.patch

cd /opt/theforeman/tfm/root/usr/share/gems/gems/foreman_bootdisk-6.1.0/

# Skip any nonexisting files
patch -p1 < /tmp/bootdisk.patch
mkdir -p public/javascripts/foreman_bootdisk/
cp app/assets/javascripts/foreman_bootdisk/host_edit.js public/javascripts/foreman_bootdisk/
```

angystardust commented:
Hi @timogoebel, thanks a lot for setting up this repo!
Today I've finally find the time to test it but i'm blocked by this strange bug when adding a vmware Computer Resource when clicking on "Load Datacenters":

> ArgumentError: wrong number of arguments(1 for 0)

I tried to test the fog connection as suggested by @shlomizadok in this gist (https://gist.github.com/shlomizadok/188696ba9d2dcb097834) and it fails with the same error so i suppose that's a bug in the fog-vsphere gem. In this hastebin link, you can see the full error: 
http://hastebin.com/wokeyegale.dos

P.s.: There's also a similar bug in redmine -> http://projects.theforeman.org/issues/10825

timogoebel commented:
@angystardust : Thanks for taking the time to test this. That you hit this bug is very unfortunate. I will try to figure out, whats going on tomorrow and get back to you.

timogoebel commented:
@angystardust : It works fine for me. Could you run the following commands from this gist to dig a little deeper and see where it blows? Thanks.

```bash
bundle
bundle exec ruby test.rb
bundle exec ruby test2.rb
```

https://gist.github.com/timogoebel/fdb5b13f5242611377fc

timogoebel commented:
@angystardust : One thing I forgot: You might want to try this patch which allows you to set the api revision passed to fog explicitly: https://patch-diff.githubusercontent.com/raw/theforeman/foreman/pull/2913.patch

angystardust commented:
Hi @timogoebel ,
I firstly tested the ruby script you provided and it seems that fog does its work and it connects to the vcenter instance.
http://hastebin.com/bapipicoxe.coffee
Next, I tried to apply the 'api version' patch but the problem is still there.
Here is a screenshot of the error:
![foreman-vsphere-debug](https://cloud.githubusercontent.com/assets/1568397/13185332/d00edc90-d740-11e5-9d8a-bd3cf2ffa54b.png)

If you want, I can try to upgrade everything to nightly.
Thanks again for your time

angystardust commented:
Now that 1.11 release was tagged it would be a pity to not include this very useful feature in.
@timogoebel , is there anything I can do to help test this?
Unlucky, I'm still blocked on the "wrong number of arguments(1 for 0)" bug :disappointed: 

@domcleal , any news about the #10825?
It seems that this one is blocking some users from using foreman with a vmware CR.

domcleal commented:
1.11 tagging doesn't matter to this PR, plugins such as bootdisk have their own release cycle.  I don't know anything about that bug, it's still marked New.

timogoebel commented:
@angystardust : Sorry, have not had any time to look at this further.

angystardust commented:
Hello again, @timogoebel !
Finally, after upgrading my test env to 1.11RC1 I was able to add the VMware CR successfuly (the "wrong argument" bug disappeared :tada: ) but I have encountered some other foreman bugs which made my test a quite painful experience. :fearful: 

**unable to connect the vNIC to the choosen DistributedVirtualPortGroup**
The first and the most bad one is a bug related to the DistributedVirtualPortGroup selection.
In the host creation phase, on the 'Interface' tab, you can choose one of the dvPortGroup exposed by the vcenter. So, I choosed the dvPortGroup related to the VLAN on which I wanted to deploy my VM. The vm is created by the fog library with all the correct settings but the vNIC always remains disconnected. 
Even trying to connect it by hand from the vsphere webclient gives a bad error: "`Cannot connect ethernet0`". 
I've tried many different options 'till I've found that the workaround is to leave the first dvPortGroup selected by foreman without modifying it (unluckly for me, the first one in the list is the dvPortGroup we use for NFS traffic and not for vms traffic), submit the host creation and than modify the dvPort from the vsphere webclient once the vm has been created and power-on.
I'll open a Redmine issue about this bug and ask @domcleal to mark it to be fixed for the 1.11 GA 'cause I think this makes fhe 'foreman+vmware' case pretty unusable.
I haven't the time to check if this bug is also present in the 1.10 branch.

**ipxe e1000 interface freeze randomly during the download of the vmlinuz and initram files (only works with vmxnet3)**
Once I've managed to connect the vNIC of the vm (with the aforementioned workaround), the ipxe bootstrap provided by the generated iso didn't worked 'cause the it was unable to fully download the vmlinuz file. I've tried to reboot many times and every time the download stopped randomly. Only once, it successfully downloaded the vmlinuz file but it suddenly stopped while downloading the initram file.
Searching on the web, I've found that this bug was fixed on the ipxe git repo with a patch submitted on the 21/04/15:
https://git.ipxe.org/ipxe.git/commit/8958f62a1c85ff6b449a685a729b61b9ea69f50a
Maybe I have to open a bugzilla to the ipxe-bootimgs mainteiner (seems that Miroslav Rezanina is my man) to cherrypick this patch and build a new rpm.
The workaround in this case was to create the vNIC as a vmxnet3 (which is also a better default for modern guest OS types) but, since the e1000 is the default nic when you create a vmware host in foreman, i think that this should be fixed.
Meanwhile, I think we should document this bug to others users who want to try this provisioning method.
@domcleal , do you think it will be useful to include this warning in the foreman manual?

In the end, after all this glitches and woes, i can tell you that your plugin do the job. :+1: 

Just 3 minor improvements and suggestions:

1) **generated isos are stored on a datastore which is selected (randomly?) the first time and it can't be changed**
We usually store all the iso files on a separate NFS datastore.
Is it possible to let the user choose the default datastore to store the iso files on (maybe selecting it from a list exposed in the settings of the Bootdisk plugin)?

2) **the isos are never deleted from the datastore**
Even if the generated isos are very small, it's not so clean to leave them around in the datastore after the build process is over. Maybe you can create a foreman-rake task or an hammer script and let it run by cron.
The workaround in this case can be to create an external script that make the cleaning job during the night time. Rvc (ruby vsphere client) is my friend. :smile: 

3) **bootdisk iso is still attached to the vm after the build is over**
As you already mentioned, the iso remains attached to the vm after the provisioning is over.
To solve this, I agree with you about the use of `reboot --eject` in the template.
Unluckily, this works in el6 but not in el7. There's a bug opened:
https://bugzilla.redhat.com/show_bug.cgi?id=1191386

Sorry for my pretty long post but that's what I mean for feedback :smile_cat:

timogoebel commented:
@angystardust: Thanks for taking the time to test this. Your feedback is very appreciated. Don't worry about the long post. Glad the patch works :-)

> unable to connect the vNIC to the choosen DistributedVirtualPortGroup

This is very odd. Let's open a redmine for this and try to investigate this.

> ipxe e1000 interface freeze randomly during the download of the vmlinuz and initram files (only works with vmxnet3)

We see that as well, but as we use "vmxnet3" everywhere, this is no problem for us. I vote for adding this to the manual.

> 1) generated isos are stored on a datastore which is selected (randomly?) the first time and it can't be changed

Actually they are stored on the datastore where the vm's first disk is created. I'd vote for merging this and add that feature in a separate PR. This one is quite large already.

> 2) the isos are never deleted from the datastore

I'd like to do this in a separate PR as well. Depending on which criterias would you delete the isos? You'd have to go through all datastores and see if there are some "old" isos lying around.

> 3) bootdisk iso is still attached to the vm after the build is over

I'll open a PR for community templates to add this behaviour. I think we cannot workaround this bug without using the kickstart functionality without calling foreman "after the first boot" to trigger the disattach action. But that is something I'd honestly rather not do.

timogoebel commented:
[test] - tests have failures, but i don't think the failures are in any way related.

angystardust commented:
Hi @timogoebel , glad you liked my review. ;)

> This is very odd. Let's open a redmine for this and try to investigate this.

I've just opened a new issue on redmine 3 days ago (http://projects.theforeman.org/issues/13932) but it was marked as a duplicate of #13281 (http://projects.theforeman.org/issues/13281). Maybe today i'm gonna find some spare time to test the attached workaround. Hope to see that fixed in time for 1.11GA release.

> We see that as well, but as we use "vmxnet3" everywhere, this is no problem for us. I vote for adding this to the manual.

I also opened a bug for this one to the ipxe mantainers (https://bugzilla.redhat.com/show_bug.cgi?id=1312754). Currently, the bug is marked as MODIFIED and "Fixed In Version" ipxe-20150821-1.git4e03af8e.el7 (rebased from fedora, i suppose) so a new rpm version should be released in RHEL/CentOS sooner or later (maybe with RHEL 7.3?).
I don't know the timing for that but maybe it's still a good idea to document this bug for poor men trying to boot from a e1000 nic on vmware.

> Actually they are stored on the datastore where the vm's first disk is created. I'd vote for merging this and add that feature in a separate PR. This one is quite large already.

I agree :+1: 
Just for my curiosity, where's this information stored? As a temporary workaround, I'm thinking about modifying the value by hand to point to a different datastore...
I know I could read the code and discover that all by myself but I'm a pretty lazy guy :)

> I'd like to do this in a separate PR as well. Depending on which criterias would you delete the isos? You'd have to go through all datastores and see if there are some "old" isos lying around.

Usually, we delete all generated isos older than 2 days with a one-liner `find` script . Isos are stored in an a specific NFS datastore which is also mounted on the cobbler node so it can write the .iso file there during the buildiso task and we can attach that in the new vm.
By the way, I also agree with you to implement this in a separate PR.

> I'll open a PR for community templates to add this behaviour. I think we cannot workaround this bug without using the kickstart functionality without calling foreman "after the first boot" to trigger the disattach action. But that is something I'd honestly rather not do.

I saw your PR, thanks. Maybe is it possible to trigger an action with foreman-hooks at the end of the build process?

timogoebel commented:
> Just for my curiosity, where's this information stored? As a temporary workaround, I'm thinking about modifying the value by hand to point to a different datastore...
> I know I could read the code and discover that all by myself but I'm a pretty lazy guy :)

The information is stored stored here: https://github.com/theforeman/foreman_bootdisk/pull/19/files#diff-df33b31b1f97dd02fc944f940295a718R29

angystardust commented:
Ok Thanks. ;)
Now the problem is: why are the tests failing? 
@domcleal , if the failing tests are unrelated to this code , can you merge it?
I've tested the patch and it works :+1:

shlomizadok commented:
@angystardust - could you please rebase? Thanks!

shlomizadok commented on app/models/concerns/foreman_bootdisk/host_ext.rb:
```diff
@@ -1,7 +1,11 @@
 module ForemanBootdisk::HostExt
-  extend ActiveSupport::Concerns
+  extend ActiveSupport::Concern
+  included do
```
Nitpick: space between `extend` and `included do`

angystardust commented:
Sorry @shlomizadok but I can't do that. @timogoebel

timogoebel commented:
Rebased and fixed the nitpick, @shlomizadok :-)

timogoebel commented:
@domcleal : Are we ready to merge this? @angystardust tested this and confirmed, it worked.

angystardust commented:
I re-confirm that the patch works great. 
Unlucky, I still can't use foreman in production because of #13281 
The issue seems to be also present in the downstream product (Satellite6): 
https://bugzilla.redhat.com/show_bug.cgi?id=1273400

@timogoebel, I'm wondering how are you workaround-ing that in you environment...

timogoebel commented:
@angystardust : I don't see that issue in our env at all. Is this only occuring with image based installs? We use PXE only. I am having trouble understanding what actions lead to this bug.

angystardust commented:
@timogoebel, I've ran into this while testing the bootdisk provisioning method (and not the image based one...I didn't tested that)

domcleal commented:
Merged as de59b7b, thanks @timogoebel!

lzap commented:
Excellent work, my only concern is this only supports Host Image while I encountered several bug reports of VMWare VMs not booting iPXE images due to some incompatibilities in drivers. I remember users need to select VMNET3 driver in order to get it working. Have you guys experienced the same? Would be nice to have a chance to attach Full Host Image, I am not sure. I don't work with VMWare hypervisors very often.

For the record, there is still a workaround for Satellite 6 and older Foreman versions with foreman_hooks plugin: https://github.com/hjensas/bootdisk_hook made by @hjensas

timogoebel commented:
@lzap : Yes, both @angystardust and I encountered the bug with ipxe and e1000 driver. The bug should be fixed in https://git.ipxe.org/ipxe.git/commit/8958f62a1c85ff6b449a685a729b61b9ea69f50a as @angystardust pointed out earlier.

I think, this can be improved in some ways. If I get around, I will open some Redmines. Attaching the full host image is a nice idea.

angystardust commented:
@timogoebel , you're faster then me! :smile: 
 @lzap , you can read my comment about the e1000 issue here ---> https://github.com/theforeman/foreman_bootdisk/pull/19#issuecomment-189899623

lzap commented:
Oh yeah, sorry I missed that. Well, that's exactly it. Just for the record (it does not belong in here), I've raised a RFE to rebase the last version in Satellite 6: https://bugzilla.redhat.com/show_bug.cgi?id=1323130

The subpackage is by the way in "optional" repo of RHEL and since this is the only dependency from there, we retag it into Satellite repositories. Therefore I also asked to pull it up into the main RHEL repository: https://bugzilla.redhat.com/show_bug.cgi?id=1323135

For the record the workaround is available as https://access.redhat.com/solutions/1990993

dbaars commented:
Hi,
I originally created one of the feature requests that related to this (http://projects.theforeman.org/issues/10926) but have been sidetracked on other work for a long time. Great to see the huge progress that has been made on this! I'd love to try and help test this if more testing is required - I am running a test Katello 2.4 instance which has Foreman 1.10.2. Is there a way I can add this plugin to this version? I'm pretty new to applying patches so may need some good steps to follow :-)

timogoebel commented:
@dbaars : Testing is always very appreciated. I don't know much about Katello to confidently answer your question. What you would need to do is: update fog and fog-vsphere, patch Foreman (patches that allow fog-vsphere to be a separate gem and add the plugin interface) and patch Foreman-Bootdisk.

dbaars commented:
Hi Tim

How would you do it if it were just Foreman? I think if you can add some detail to what you mapped out I'll be able to sort it for katello

Thanks heaps
Dylan

On 5/04/2016, at 6:46 PM, Timo Goebel <notifications@github.com<mailto:notifications@github.com>> wrote:


@dbaars<https://github.com/dbaars> : Testing is always very appreciated. I don't know much about Katello to confidently answer your question. What you would need to do is: update fog and fog-vsphere, patch Foreman (patches that allow fog-vsphere to be a separate gem and add the plugin interface) and patch Foreman-Bootdisk.

—
You are receiving this because you were mentioned.
Reply to this email directly or view it on GitHub<https://github.com/theforeman/foreman_bootdisk/pull/19#issuecomment-205677890>

Dylan Baars
Systems Engineer - IT Projects

+64-7-859-1829 | +64-27-412-4360 | Gate 10 Silverdale Road, Hillcrest, Hamilton | www.niwa.co.nz<http://www.niwa.co.nz>
[NIWA]<http://www.niwa.co.nz>
To ensure compliance with legal requirements and to maintain cyber security standards, NIWA's IT systems are subject to ongoing monitoring, activity logging and auditing. This monitoring and auditing service may be provided by third parties. Such third parties can access information transmitted to, processed by and stored on NIWA's IT systems.

angystardust commented:
Hi @dbaars , you can find the instructions to patch your 10.x setup in this comment: https://github.com/theforeman/foreman_bootdisk/pull/19#issuecomment-182374236

@domcleal , what about making a new release of this plugin so it can be easily installed via repository?

dbaars commented:
Hi again,
if I try and use the RPMs by adding https://tgoebel.de/trunk/tfm-packages/ as a repo I get the following if I try and do a 'yum update'

Error: Package: foreman-compute-1.10.3-1.el7.noarch (@foreman)
           Requires: tfm-rubygem(fog-core) = 1.32.1
           Removing: tfm-rubygem-fog-core-1.32.1-2.el7.noarch (@foreman)
               tfm-rubygem(fog-core) = 1.32.1
           Updated By: tfm-rubygem-fog-core-1.35.0-1.el7.noarch (tgoebel-repo)
               tfm-rubygem(fog-core) = 1.35.0
Error: Package: foreman-compute-1.10.3-1.el7.noarch (@foreman)
           Requires: tfm-rubygem(fog) = 1.34.0
           Removing: tfm-rubygem-fog-1.34.0-2.el7.noarch (@foreman)
               tfm-rubygem(fog) = 1.34.0
           Updated By: tfm-rubygem-fog-1.37.0-1.el7.noarch (tgoebel-repo)
               tfm-rubygem(fog) = 1.37.0
           Available: tfm-rubygem-fog-1.34.0-1.el7.noarch (foreman)
               tfm-rubygem(fog) = 1.34.0
 You could try using --skip-broken to work around the problem
 You could try running: rpm -Va --nofiles --nodigest

do I skip-broken? Or something else?

Thanks,
Dylan

domcleal commented:
> @domcleal , what about making a new release of this plugin so it can be easily installed via repository?

Yes, I'll be doing that shortly, it's currently being translated.

timogoebel commented:
@dbaars: The Packages won't work as there are "newer" packages in the foreman repo. Unfortunately I don't have time to build new packages right now. I should probably take them offline anyways.

The required changes for this to work were iirc:

http://projects.theforeman.org/issues/12624
http://projects.theforeman.org/issues/12612

I have described how to patch foreman-bookdisk plugin above.

domcleal commented:
Version 7.0.0's released and packages should be available in the Foreman plugin repos for 1.11 and nightly in the next 24 hours or so.  The plugin's not compatible with 1.10 without further patching.

dbaars commented:
Thanks @domcleal ! I am trying to get my test katello instance up to v3.0RC (or nightly, just any that works really) as I see the 3.0RC uses the 1.11 repos that have the version 7 RPMs for this plugin, and am struggling - https://groups.google.com/forum/#!topic/foreman-users/PUl7fr9ZJSM

Hopefully I can get some help there and then can test this plugin :-)

dbaars commented:
Hi,
I have 3.0RC of katello installed (which uses Foreman 1.11), with 7.0 RPM installed for bootdisk -

[root@wellkatellotst foreman]# yum list installed | grep bootdisk
tfm-rubygem-foreman_bootdisk.noarch    7.0.0-1.fm1_11.el7       @foreman-plugins
tfm-rubygem-hammer_cli_foreman_bootdisk.noarch
[root@wellkatellotst foreman]#

when I start creating a new host, and browse to the Operating Systems tab, I see the 3 provisioning methods available. The default option is "Network based" - if I change it to "Bootdisk Based" no options change - is that correct? As per screenshot...

![image](https://cloud.githubusercontent.com/assets/14931710/14415735/7436ef8c-fff6-11e5-8f86-3e3e9e7a09bc.png)

If I click resolve on templates (on the operating system page) is resolves to some PXE defaults which I'd guess is wrong!

If I continue and fill in all the details for the VM, and hit submit, it tries to do something but then the page refreshes with red lettering for "Operating System" - if I browse to it that page now looks like this

![image](https://cloud.githubusercontent.com/assets/14931710/14415806/23942ce2-fff7-11e5-9f19-bca94e863345.png)

I'm not sure what logs will be useful here (if any). As far as foreman packages installed:

[root@wellkatellotst foreman]# yum list installed | grep foreman
foreman.noarch                         1.11.0-1.el7             @foreman
foreman-compute.noarch                 1.11.0-1.el7             @foreman
foreman-debug.noarch                   1.11.0-1.el7             @foreman
foreman-installer.noarch               1:1.11.0-1.el7           @foreman
foreman-installer-katello.noarch       3.0.0-3.rc1.el7          @katello
foreman-postgresql.noarch              1.11.0-1.el7             @foreman
foreman-proxy.noarch                   1.11.0-1.el7             @foreman
foreman-release.noarch                 1.11.0-1.el7             @/foreman-release
foreman-release-scl.x86_64             2-1.el7                  @foreman
foreman-selinux.noarch                 1.11.0-1.el7             @foreman
foreman-vmware.noarch                  1.11.0-1.el7             @foreman

timogoebel commented:
@dbaars: Thanks for testing.

> If I click resolve on templates (on the operating system page) is resolves to some PXE defaults which I'd guess is wrong!

Yes. It should show a "Bootdisk Template". This has to be configured. Please see https://github.com/theforeman/foreman_bootdisk#templates .


> If I continue and fill in all the details for the VM, and hit submit, it tries to do something but then the page refreshes with red lettering for "Operating System"

This indeed looks like a UI bug hiding the form validation error, which can easily be fixed though. I'll prepare a patch. As a workaround you can select "Network Based" as a provision method and then select "Bootdisk Based" again. Then it should show you the error. What does it say?

dbaars commented:
Hi @timogoebel 

if I change between the different provisioning methods nothing happens (as in no error is displayed) - the options certainly change between image based and network based, just if I go to bootdisk based nothing happens. No error or otherwise in the production.log either for foreman

timogoebel commented:
@dbaars : What happens if you run "bootdisk_provision_method_selected();" on the Java Script console of your browser?

dbaars commented:
I get this

VM77:2 Uncaught ReferenceError: bootdisk_provision_method_selected is not defined(…)(anonymous function) @ VM77:2InjectedScript._evaluateOn @ VM71:878InjectedScript._evaluateAndWrap @ VM71:811InjectedScript.evaluate @ VM71:667

dbaars commented:
But there is also a whole bunch of other errors even before I get to the operating systems tab 

![image](https://cloud.githubusercontent.com/assets/14931710/14454183/3447f4d2-00ed-11e6-9957-9ba0f7c23e9f.png)

timogoebel commented:
@dbaars : It does not load the foreman_bootdisk/host_edit.js which results in the UI error. I have no idea though, why it throws a 404.
Could you check that the file exists on your os? It should be located here: ```/opt/theforeman/tfm/root/usr/share/gems/gems/foreman_bootdisk-7.0.0/app/assets/javascripts/foreman_bootdisk/host_edit.js```

dbaars commented:
[root@wellkatellotst foreman_bootdisk]# pwd
/opt/theforeman/tfm/root/usr/share/gems/gems/foreman_bootdisk-7.0.0/app/assets/javascripts/foreman_bootdisk
[root@wellkatellotst foreman_bootdisk]# ls -l
total 4
-rw-r--r--. 1 root root 315 Apr  6 23:30 host_edit.js

dbaars commented:
I don't know if this helps, but when I first browse to the "new hosts" page I get this in /var/log/foreman/production.log (among other things obviously)

2016-04-13 12:35:30 [app] [I] Started GET "/javascripts/foreman_bootdisk/host_edit.js" for 192.168.222.132 at 2016-04-13 12:35:30 +1200
2016-04-13 12:35:30 [app] [F]
 | ActionController::RoutingError (No route matches [GET] "/javascripts/foreman_bootdisk/host_edit.js"):
 |   actionpack (4.1.5) lib/action_dispatch/middleware/debug_exceptions.rb:21:in `call'
 |   actionpack (4.1.5) lib/action_dispatch/middleware/show_exceptions.rb:30:in `call'
 |   railties (4.1.5) lib/rails/rack/logger.rb:38:in `call_app'
 |   railties (4.1.5) lib/rails/rack/logger.rb:22:in `call'
 |   actionpack (4.1.5) lib/action_dispatch/middleware/request_id.rb:21:in `call'
 |   rack (1.5.2) lib/rack/methodoverride.rb:21:in `call'
 |   rack (1.5.2) lib/rack/runtime.rb:17:in `call'
 |   activesupport (4.1.5) lib/active_support/cache/strategy/local_cache_middleware.rb:26:in `call'
 |   actionpack (4.1.5) lib/action_dispatch/middleware/static.rb:64:in `call'
 |   rack (1.5.2) lib/rack/sendfile.rb:112:in `call'
 |   railties (4.1.5) lib/rails/engine.rb:514:in `call'
 |   railties (4.1.5) lib/rails/application.rb:144:in `call'
 |   railties (4.1.5) lib/rails/railtie.rb:194:in `public_send'
 |   railties (4.1.5) lib/rails/railtie.rb:194:in `method_missing'
 |   rack (1.5.2) lib/rack/builder.rb:138:in `call'
 |   rack (1.5.2) lib/rack/urlmap.rb:65:in `block in call'
 |   rack (1.5.2) lib/rack/urlmap.rb:50:in `each'
 |   rack (1.5.2) lib/rack/urlmap.rb:50:in `call'
 |   /usr/share/gems/gems/passenger-4.0.18/lib/phusion_passenger/rack/thread_handler_extension.rb:77:in `process_request'
 |   /usr/share/gems/gems/passenger-4.0.18/lib/phusion_passenger/request_handler/thread_handler.rb:140:in `accept_and_process_next_request'
 |   /usr/share/gems/gems/passenger-4.0.18/lib/phusion_passenger/request_handler/thread_handler.rb:108:in `main_loop'
 |   /usr/share/gems/gems/passenger-4.0.18/lib/phusion_passenger/request_handler.rb:441:in `block (3 levels) in start_threads'
 |   logging (1.8.2) lib/logging/diagnostic_context.rb:323:in `call'
 |   logging (1.8.2) lib/logging/diagnostic_context.rb:323:in `block in create_with_logging_context'
 |
 |

timogoebel commented:
@dbaars: Thanks for bearing with me here. I had a look into the issue and nailed it down to a packaging problem (the compiled file is simply missing). I've already opened a PR to get this fixed. I'll let you know when you can retest this.

dbaars commented:
@timogoebel - magic! Thanks heaps :-)

timogoebel commented:
@dbaars : I just tested the newly build rpms with Foreman 1.11 and they work now. Can you give this another shot?
```
yum clean all
yum upgrade tfm-rubygem-foreman_bootdisk
systemctl restart httpd.service
```

dbaars commented:
Morning @timogoebel 
when I browse to the operating system tab I no longer get any errors, but selecting "boot disk based" seems to give me the "network based" options - and if I open the javascript console and put in "bootdisk_provision_method_selected();" I get 

```
bootdisk_provision_method_selected();
undefined
```

The file is there...
```
[root@wellkatellotst foreman_bootdisk]# ls
host_edit.js
[root@wellkatellotst foreman_bootdisk]# ls -l
total 4
-rw-r--r--. 1 root root 315 Apr 19 03:22 host_edit.js
```

correct version I think?
```
[root@wellkatellotst foreman_bootdisk]# cat /var/log/yum.log | grep bootdisk-7.0.0-2
Apr 20 06:14:42 Updated: tfm-rubygem-foreman_bootdisk-7.0.0-2.fm1_11.el7.noarch
[root@wellkatellotst foreman_bootdisk]#
[root@wellkatellotst foreman_bootdisk]# yum list installed | grep bootdisk
tfm-rubygem-foreman_bootdisk.noarch    7.0.0-2.fm1_11.el7       @foreman-plugins
tfm-rubygem-hammer_cli_foreman_bootdisk.noarch
[root@wellkatellotst foreman_bootdisk]#
```

Nothing in the production.log that I can see. Thoughts? Sorry :(

timogoebel commented:
@dbaars : This is fine and expected behaviour. Bootdisk and Network Based installs require the same settings, that's why the ui does not change. The ISO just replaces the PXE part of the installation in order to load the actual os installer (anaconda for example).

dbaars commented:
OK, so I tried it using as much "default" stuff as possible - here's what I have selected/resolved on the operating system tab

![image](https://cloud.githubusercontent.com/assets/14931710/14658799/ae813b34-06ea-11e6-8c73-dbad64106a70.png)

Note that when I'm selecting lifecycle environment/content view I get the following errors in the javascript console

![image](https://cloud.githubusercontent.com/assets/14931710/14658913/91e83da0-06eb-11e6-93e3-0e70509813b4.png)

when I click submit to create the host I don't see anything in the GUI but the following in the production.log

```
2016-04-20 11:21:10 [app] [I]   Rendered common_parameters/_inherited_parameters.html.erb (9.8ms)
2016-04-20 11:21:10 [app] [I] Completed 200 OK in 278ms (Views: 15.7ms | ActiveRecord: 30.7ms)
2016-04-20 11:21:10 [app] [I] Processing by ImagesController#index as JSON
2016-04-20 11:21:10 [app] [I]   Parameters: {"search"=>"operatingsystem=2 architecture=1", "compute_resource_id"=>"1"}
2016-04-20 11:21:10 [app] [I] Completed 200 OK in 192ms (Views: 9.8ms | ActiveRecord: 21.7ms)
2016-04-20 11:21:18 [app] [I] Started POST "/hosts/template_used?provisioning=bootdisk" for 192.168.222.132 at 2016-04-20 11:21:18 +1200
2016-04-20 11:21:18 [app] [I] Processing by HostsController#template_used as */*
2016-04-20 11:21:18 [app] [I]   Parameters: {"utf8"=>"✓", "authenticity_token"=>"3IRb/egtU+GCm7prejLGK/2qckj+hTt1PFXJG/nJS98=", "host"=>{"salt_module_ids"=>[""], "name"=>"testvm", "organization_id"=>"1", "location_id"=>"16", "hostgroup_id"=>"3", "compute_resource_id"=>"1", "compute_profile_id"=>"3", "content_facet_attributes"=>{"lifecycle_environment_id"=>"18", "content_view_id"=>"7", "id"=>""}, "environment_id"=>"", "content_source_id"=>"1", "salt_environment_id"=>"2", "salt_proxy_id"=>"1", "managed"=>"true", "progress_report_id"=>"[FILTERED]", "type"=>"Host::Managed", "interfaces_attributes"=>{"0"=>{"_destroy"=>"0", "type"=>"Nic::Managed", "mac"=>"", "identifier"=>"", "name"=>"testvm", "domain_id"=>"1", "subnet_id"=>"2", "ip"=>"192.168.14.120", "managed"=>"0", "primary"=>"1", "provision"=>"1", "virtual"=>"0", "tag"=>"", "attached_to"=>"", "compute_attributes"=>{"type"=>"VirtualVmxnet3", "network"=>"SDT Network"}}, "new_interfaces"=>{"_destroy"=>"1", "type"=>"Nic::Managed", "mac"=>"", "identifier"=>"", "name"=>"", "domain_id"=>"", "subnet_id"=>"", "ip"=>"", "managed"=>"1", "primary"=>"0", "provision"=>"0", "virtual"=>"0", "tag"=>"", "attached_to"=>"", "compute_attributes"=>{"type"=>"VirtualE1000", "network"=>"Allen User Network"}}}, "compute_attributes"=>{"cpus"=>"4", "corespersocket"=>"1", "memory_mb"=>"4096", "cluster"=>"wellpool05", "resource_pool"=>"Resources", "path"=>"/Datacenters/Wellington/vm", "guest_id"=>"centos64Guest", "scsi_controller_type"=>"VirtualLsiLogicController", "hardware_version"=>"vmx-10", "memoryHotAddEnabled"=>"0", "cpuHotAddEnabled"=>"0", "start"=>"1", "volumes_attributes"=>{"new_volumes"=>{"_delete"=>"", "datastore"=>"SMT ISO NFS", "name"=>"Hard disk", "size_gb"=>"10", "thin"=>"true", "eager_zero"=>"false"}, "0"=>{"_delete"=>"", "datastore"=>"wellpool05 SR1", "name"=>"Hard disk", "size_gb"=>"30", "thin"=>"false", "eager_zero"=>"false"}}}, "architecture_id"=>"1", "operatingsystem_id"=>"2", "provision_method"=>"bootdisk", "build"=>"1", "medium_id"=>"", "ptable_id"=>"61", "disk"=>"", "root_pass"=>"[FILTERED]", "is_owned_by"=>"3-Users", "enabled"=>"1", "comment"=>"", "overwrite"=>"false"}, "capabilities"=>"build image bootdisk", "provider"=>"Vmware", "fakepassword"=>"[FILTERED]", "bare_metal_capabilities"=>"build", "provisioning"=>"bootdisk"}
2016-04-20 11:21:18 [app] [I]   Rendered hosts/_provisioning.html.erb (11.3ms)
2016-04-20 11:21:18 [app] [I] Completed 200 OK in 349ms (Views: 15.1ms | ActiveRecord: 41.1ms)

2016-04-20 11:21:54 [app] [I] Started POST "/hosts" for 192.168.222.132 at 2016-04-20 11:21:54 +1200
2016-04-20 11:21:54 [app] [I] Processing by HostsController#create as */*
2016-04-20 11:21:54 [app] [I]   Parameters: {"utf8"=>"✓", "authenticity_token"=>"3IRb/egtU+GCm7prejLGK/2qckj+hTt1PFXJG/nJS98=", "host"=>{"salt_module_ids"=>[""], "name"=>"testvm", "organization_id"=>"1", "location_id"=>"16", "hostgroup_id"=>"3", "compute_resource_id"=>"1", "compute_profile_id"=>"3", "content_facet_attributes"=>{"lifecycle_environment_id"=>"18", "content_view_id"=>"7", "id"=>""}, "environment_id"=>"", "content_source_id"=>"1", "salt_environment_id"=>"2", "salt_proxy_id"=>"1", "managed"=>"true", "progress_report_id"=>"[FILTERED]", "type"=>"Host::Managed", "interfaces_attributes"=>{"0"=>{"_destroy"=>"0", "type"=>"Nic::Managed", "mac"=>"", "identifier"=>"", "name"=>"testvm", "domain_id"=>"1", "subnet_id"=>"2", "ip"=>"192.168.14.120", "managed"=>"0", "primary"=>"1", "provision"=>"1", "virtual"=>"0", "tag"=>"", "attached_to"=>"", "compute_attributes"=>{"type"=>"VirtualVmxnet3", "network"=>"SDT Network"}}}, "compute_attributes"=>{"cpus"=>"4", "corespersocket"=>"1", "memory_mb"=>"4096", "cluster"=>"wellpool05", "resource_pool"=>"Resources", "path"=>"/Datacenters/Wellington/vm", "guest_id"=>"centos64Guest", "scsi_controller_type"=>"VirtualLsiLogicController", "hardware_version"=>"vmx-10", "memoryHotAddEnabled"=>"0", "cpuHotAddEnabled"=>"0", "start"=>"1", "volumes_attributes"=>{"0"=>{"_delete"=>"", "datastore"=>"wellpool05 SR1", "name"=>"Hard disk", "size_gb"=>"30", "thin"=>"false", "eager_zero"=>"false"}}}, "architecture_id"=>"1", "operatingsystem_id"=>"2", "provision_method"=>"bootdisk", "build"=>"1", "medium_id"=>"", "ptable_id"=>"61", "disk"=>"", "root_pass"=>"[FILTERED]", "is_owned_by"=>"3-Users", "enabled"=>"1", "comment"=>"", "overwrite"=>"false"}, "capabilities"=>"build image bootdisk", "provider"=>"Vmware", "bare_metal_capabilities"=>"build"}
2016-04-20 11:21:55 [app] [I] Failed to save: Subnet is not defined for host's organization., Medium can't be blank
2016-04-20 11:21:55 [app] [I] Deface: [WARNING] No :original defined for 'hosts_update_environments_select', you should change its definition to include:
 |  :original => 'ec4418ddc20ff0c3667c8361538717db7690a13a'
2016-04-20 11:21:55 [app] [I] Deface: [WARNING] No :original defined for 'add_salt_environment_to_host', you should change its definition to include:
 |  :original => '529fa978c698865a26cfd4ee00c810cc00e341cb'
2016-04-20 11:21:55 [app] [I] Deface: [WARNING] No :original defined for 'add_salt_modules_tab_to_host', you should change its definition to include:
 |  :original => 'f234dfccee67a10906661b649c8cf0ee976e3c58'
2016-04-20 11:21:55 [app] [I] Deface: [WARNING] No :original defined for 'add_salt_modules_tab_pane_to_host', you should change its definition to include:
 |  :original => '6826cab75371ce22dda4ea0c899e6a3ea2caa740'
2016-04-20 11:21:55 [app] [I] Deface: [WARNING] No :original defined for 'add_salt_proxy_to_host', you should change its definition to include:
 |  :original => '6826cab75371ce22dda4ea0c899e6a3ea2caa740'
2016-04-20 11:21:55 [app] [I]   Rendered hosts/_progress.html.erb (2.0ms)
2016-04-20 11:21:56 [app] [I] Started GET "/tasks/05a525be-7c54-4b47-8775-ba31162c20b6" for 192.168.222.132 at 2016-04-20 11:21:56 +1200
2016-04-20 11:21:56 [app] [I] Processing by TasksController#show as */*
2016-04-20 11:21:56 [app] [I]   Parameters: {"id"=>"05a525be-7c54-4b47-8775-ba31162c20b6"}
2016-04-20 11:21:56 [app] [I]   Rendered tasks/_list.html.erb (1.1ms)
2016-04-20 11:21:56 [app] [I] Completed 200 OK in 20ms (Views: 6.5ms | ActiveRecord: 1.7ms)
2016-04-20 11:21:56 [app] [I]   Rendered nic/_base_form.html.erb (64.6ms)
2016-04-20 11:21:56 [app] [I]   Rendered nic/_virtual_form.html.erb (5.5ms)
2016-04-20 11:21:57 [app] [I]   Rendered compute_resources_vms/form/vmware/_network.html.erb (1267.0ms)
2016-04-20 11:21:57 [app] [I]   Rendered nic/_provider_specific_form.html.erb (1276.3ms)
2016-04-20 11:21:57 [app] [I]   Rendered nic/manageds/_managed.html.erb (1373.8ms)
2016-04-20 11:21:57 [app] [I]   Rendered nic/_base_form.html.erb (54.2ms)
2016-04-20 11:21:57 [app] [I]   Rendered nic/_virtual_form.html.erb (3.8ms)
2016-04-20 11:21:57 [app] [I] Started GET "/tasks/05a525be-7c54-4b47-8775-ba31162c20b6" for 192.168.222.132 at 2016-04-20 11:21:57 +1200
2016-04-20 11:21:57 [app] [I] Processing by TasksController#show as */*
2016-04-20 11:21:57 [app] [I]   Parameters: {"id"=>"05a525be-7c54-4b47-8775-ba31162c20b6"}
2016-04-20 11:21:57 [app] [I]   Rendered tasks/_list.html.erb (0.3ms)
2016-04-20 11:21:57 [app] [I] Completed 200 OK in 16ms (Views: 1.8ms | ActiveRecord: 2.2ms)
2016-04-20 11:21:58 [app] [I]   Rendered compute_resources_vms/form/vmware/_network.html.erb (419.2ms)
2016-04-20 11:21:58 [app] [I]   Rendered nic/_provider_specific_form.html.erb (421.4ms)
2016-04-20 11:21:58 [app] [I]   Rendered nic/manageds/_managed.html.erb (494.2ms)
2016-04-20 11:21:58 [app] [I]   Rendered hosts/_interfaces.html.erb (1892.0ms)
2016-04-20 11:21:59 [app] [I]   Rendered compute_resources_vms/form/vmware/_base.html.erb (1400.3ms)
2016-04-20 11:21:59 [app] [I] Started GET "/tasks/05a525be-7c54-4b47-8775-ba31162c20b6" for 192.168.222.132 at 2016-04-20 11:21:59 +1200
2016-04-20 11:21:59 [app] [I] Processing by TasksController#show as */*
2016-04-20 11:21:59 [app] [I]   Parameters: {"id"=>"05a525be-7c54-4b47-8775-ba31162c20b6"}
2016-04-20 11:21:59 [app] [I]   Rendered tasks/_list.html.erb (0.2ms)
2016-04-20 11:21:59 [app] [I] Completed 200 OK in 10ms (Views: 0.9ms | ActiveRecord: 1.4ms)
2016-04-20 11:22:01 [app] [I] Started GET "/tasks/05a525be-7c54-4b47-8775-ba31162c20b6" for 192.168.222.132 at 2016-04-20 11:22:01 +1200
2016-04-20 11:22:01 [app] [I] Processing by TasksController#show as */*
2016-04-20 11:22:01 [app] [I]   Parameters: {"id"=>"05a525be-7c54-4b47-8775-ba31162c20b6"}
2016-04-20 11:22:01 [app] [I]   Rendered tasks/_list.html.erb (0.2ms)
2016-04-20 11:22:01 [app] [I] Completed 200 OK in 14ms (Views: 1.2ms | ActiveRecord: 2.3ms)
2016-04-20 11:22:02 [app] [I]   Rendered compute_resources_vms/form/vmware/_volume.html.erb (3271.6ms)
2016-04-20 11:22:02 [app] [I] Started GET "/tasks/05a525be-7c54-4b47-8775-ba31162c20b6" for 192.168.222.132 at 2016-04-20 11:22:02 +1200
2016-04-20 11:22:02 [app] [I] Processing by TasksController#show as */*
2016-04-20 11:22:02 [app] [I]   Parameters: {"id"=>"05a525be-7c54-4b47-8775-ba31162c20b6"}
2016-04-20 11:22:02 [app] [I]   Rendered tasks/_list.html.erb (0.1ms)
2016-04-20 11:22:02 [app] [I] Completed 200 OK in 10ms (Views: 0.8ms | ActiveRecord: 1.3ms)
2016-04-20 11:22:04 [app] [I] Started GET "/tasks/05a525be-7c54-4b47-8775-ba31162c20b6" for 192.168.222.132 at 2016-04-20 11:22:04 +1200
2016-04-20 11:22:04 [app] [I] Processing by TasksController#show as */*
2016-04-20 11:22:04 [app] [I]   Parameters: {"id"=>"05a525be-7c54-4b47-8775-ba31162c20b6"}
2016-04-20 11:22:04 [app] [I]   Rendered tasks/_list.html.erb (0.2ms)
2016-04-20 11:22:04 [app] [I] Completed 200 OK in 14ms (Views: 1.1ms | ActiveRecord: 2.3ms)
2016-04-20 11:22:06 [app] [I] Started GET "/tasks/05a525be-7c54-4b47-8775-ba31162c20b6" for 192.168.222.132 at 2016-04-20 11:22:06 +1200
2016-04-20 11:22:06 [app] [I] Processing by TasksController#show as */*
2016-04-20 11:22:06 [app] [I]   Parameters: {"id"=>"05a525be-7c54-4b47-8775-ba31162c20b6"}
2016-04-20 11:22:06 [app] [I]   Rendered tasks/_list.html.erb (0.4ms)
2016-04-20 11:22:06 [app] [I] Completed 200 OK in 12ms (Views: 1.6ms | ActiveRecord: 1.7ms)
2016-04-20 11:22:06 [app] [I]   Rendered compute_resources_vms/form/vmware/_volume.html.erb (3908.6ms)
2016-04-20 11:22:06 [app] [I]   Rendered compute_resources_vms/form/_volumes.html.erb (7189.2ms)
2016-04-20 11:22:06 [app] [I]   Rendered hosts/_compute_detail.html.erb (8607.1ms)
2016-04-20 11:22:06 [app] [I]   Rendered hosts/_compute.html.erb (8625.7ms)
2016-04-20 11:22:06 [app] [I]   Rendered common/os_selection/_architecture.html.erb (23.3ms)
2016-04-20 11:22:06 [app] [I]   Rendered /opt/theforeman/tfm/root/usr/share/gems/gems/foreman_bootdisk-7.0.0/app/views/hosts/provision_method/bootdisk/_form.html.erb (1.2ms)
2016-04-20 11:22:06 [app] [I]   Rendered common/os_selection/_operatingsystem.html.erb (140.3ms)
2016-04-20 11:22:06 [app] [I]   Rendered hosts/provision_method/build/_form.html.erb (147.7ms)
2016-04-20 11:22:06 [app] [I]   Rendered hosts/provision_method/image/_form.html.erb (0.9ms)
2016-04-20 11:22:06 [app] [I]   Rendered hosts/_operating_system.html.erb (205.0ms)
2016-04-20 11:22:06 [app] [I]   Rendered hosts/_unattended.html.erb (8843.5ms)
2016-04-20 11:22:06 [app] [I]   Rendered puppetclasses/_class_parameters.html.erb (0.0ms)
2016-04-20 11:22:06 [app] [I]   Rendered puppetclasses/_classes_parameters.html.erb (58.2ms)
2016-04-20 11:22:07 [app] [I]   Rendered common_parameters/_inherited_parameters.html.erb (8.9ms)
2016-04-20 11:22:07 [app] [I]   Rendered common_parameters/_parameter.html.erb (4.8ms)
2016-04-20 11:22:07 [app] [I]   Rendered common_parameters/_parameters.html.erb (24.9ms)
2016-04-20 11:22:07 [app] [I]   Rendered hosts/_form.html.erb (11202.0ms)
2016-04-20 11:22:07 [app] [I]   Rendered hosts/new.html.erb within layouts/application (11288.4ms)
2016-04-20 11:22:07 [app] [I]   Rendered layouts/_application_content.html.erb (1.4ms)
2016-04-20 11:22:07 [app] [I]   Rendered home/_user_dropdown.html.erb (29.9ms)
2016-04-20 11:22:07 [app] [I] Read fragment views/tabs_and_title_records-3 (0.3ms)
2016-04-20 11:22:07 [app] [I]   Rendered home/_topbar.html.erb (37.3ms)
2016-04-20 11:22:07 [app] [I]   Rendered layouts/base.html.erb (42.8ms)
2016-04-20 11:22:07 [app] [I] Completed 200 OK in 12558ms (Views: 11266.3ms | ActiveRecord: 144.0ms)
2016-04-20 11:22:07 [app] [I] Started PUT "/salt/minions/salt_environment_selected" for 192.168.222.132 at 2016-04-20 11:22:07 +1200
2016-04-20 11:22:07 [app] [I] Started GET "/operatingsystems/2/available_kickstart_repo?lifecycle_environment_id=18&content_source_id=1&architecture_id=1&operatingsystem_id=2&content_view_id=7" for 192.168.222.132 at 2016-04-20 11:22:07 +1200
2016-04-20 11:22:07 [app] [I] Processing by OperatingsystemsController#available_kickstart_repo as */*
2016-04-20 11:22:07 [app] [I]   Parameters: {"lifecycle_environment_id"=>"18", "content_source_id"=>"1", "architecture_id"=>"1", "operatingsystem_id"=>"2", "content_view_id"=>"7", "id"=>"2"}
2016-04-20 11:22:07 [app] [I] Processing by ForemanSalt::MinionsController#salt_environment_selected as */*
2016-04-20 11:22:07 [app] [I]   Parameters: {"utf8"=>"✓", "authenticity_token"=>"3IRb/egtU+GCm7prejLGK/2qckj+hTt1PFXJG/nJS98=", "host"=>{"salt_module_ids"=>[""], "name"=>"testvm", "organization_id"=>"1", "location_id"=>"16", "hostgroup_id"=>"3", "compute_resource_id"=>"1", "compute_profile_id"=>"3", "content_facet_attributes"=>{"lifecycle_environment_id"=>"18", "content_view_id"=>"7", "id"=>""}, "environment_id"=>"", "content_source_id"=>"1", "salt_environment_id"=>"2", "salt_proxy_id"=>"1", "managed"=>"true", "progress_report_id"=>"[FILTERED]", "type"=>"Host::Managed", "interfaces_attributes"=>{"0"=>{"_destroy"=>"0", "type"=>"Nic::Managed", "mac"=>"", "identifier"=>"", "name"=>"testvm", "domain_id"=>"1", "subnet_id"=>"2", "ip"=>"192.168.14.120", "managed"=>"0", "primary"=>"1", "provision"=>"1", "virtual"=>"0", "tag"=>"", "attached_to"=>"", "compute_attributes"=>{"type"=>"VirtualVmxnet3", "network"=>"SDT Network"}}, "new_interfaces"=>{"_destroy"=>"1", "type"=>"Nic::Managed", "mac"=>"", "identifier"=>"", "name"=>"", "domain_id"=>"", "subnet_id"=>"", "ip"=>"", "managed"=>"1", "primary"=>"0", "provision"=>"0", "virtual"=>"0", "tag"=>"", "attached_to"=>"", "compute_attributes"=>{"type"=>"VirtualE1000", "network"=>"Allen User Network"}}}, "compute_attributes"=>{"cpus"=>"4", "corespersocket"=>"1", "memory_mb"=>"4096", "cluster"=>"wellpool05", "resource_pool"=>"Resources", "path"=>"/Datacenters/Wellington/vm", "guest_id"=>"centos64Guest", "scsi_controller_type"=>"VirtualLsiLogicController", "hardware_version"=>"vmx-10", "memoryHotAddEnabled"=>"0", "cpuHotAddEnabled"=>"0", "start"=>"1", "volumes_attributes"=>{"new_volumes"=>{"_delete"=>"", "datastore"=>"SMT ISO NFS", "name"=>"Hard disk", "size_gb"=>"10", "thin"=>"true", "eager_zero"=>"false"}, "0"=>{"_delete"=>"", "datastore"=>"wellpool05 SR1", "name"=>"Hard disk", "size_gb"=>"30", "thin"=>"false", "eager_zero"=>"false"}}}, "architecture_id"=>"1", "operatingsystem_id"=>"2", "provision_method"=>"bootdisk", "build"=>"1", "medium_id"=>"", "ptable_id"=>"61", "disk"=>"", "root_pass"=>"[FILTERED]", "is_owned_by"=>"3-Users", "enabled"=>"1", "comment"=>"", "overwrite"=>"false"}, "capabilities"=>"build image bootdisk", "provider"=>"Vmware", "fakepassword"=>"[FILTERED]", "bare_metal_capabilities"=>"build", "host_id"=>"null"}
2016-04-20 11:22:07 [app] [I] Completed 200 OK in 53ms (Views: 0.6ms | ActiveRecord: 9.4ms)
2016-04-20 11:22:07 [app] [I]   Rendered /opt/theforeman/tfm/root/usr/share/gems/gems/foreman_salt-5.0.0/app/views/foreman_salt/salt_modules/_host_tab_pane.html.erb (13.3ms)
2016-04-20 11:22:07 [app] [I] Completed 200 OK in 136ms (Views: 26.9ms | ActiveRecord: 20.5ms)
2016-04-20 11:22:07 [app] [I] Started GET "/tasks/05a525be-7c54-4b47-8775-ba31162c20b6" for 192.168.222.132 at 2016-04-20 11:22:07 +1200
2016-04-20 11:22:07 [app] [I] Processing by TasksController#show as */*
2016-04-20 11:22:07 [app] [I]   Parameters: {"id"=>"05a525be-7c54-4b47-8775-ba31162c20b6"}
2016-04-20 11:22:07 [app] [I]   Rendered tasks/_list.html.erb (1.0ms)
2016-04-20 11:22:07 [app] [I] Completed 200 OK in 20ms (Views: 6.3ms | ActiveRecord: 2.3ms)
```

this is what I see on the operating system tab after the above is done 

![image](https://cloud.githubusercontent.com/assets/14931710/14658770/86331fc6-06ea-11e6-9625-4ff6ef659571.png)

nothing happens in VMware

Help :-)

dbaars commented:
Hi @timogoebel , just checking in - can I provide any further information to help troubleshooting?

dbaars commented:
Hi again @timogoebel. I have a second dev server which was running Katello 2.4, with the release of Katello 3.0RC3 I attempted the upgrade and it succeeded. I then tried a bootdisk based install on this server and it worked!!! I did see these errors in the javascript console

![image](https://cloud.githubusercontent.com/assets/14931710/14733359/86a0dda4-08b6-11e6-80a0-a4f56afe1406.png)

and the default kickstart files didn't seem to install katello-agent or salt-minion (I think because of lack of appropriate repositories), BUT the boot disk based install certainly worked, now I have to learn a bit about foreman provisioning to get our VMs installed as we would like. 

Such an AWESOME feature to add, thanks so much for your hard work :-) 

I'm more than happy to continue troubleshooting the other servers issues if you want, but it might just be a katello 3.0RC2 issue that isn't worth pursuing any further? (note I also upgraded that server to 3.0RC3 and still see the same problem, but if you think worthwhile I can rebuild that VM and install 3.0RC3 from scratch and see if the problems persist? Or just bin it as a weird one-off problem)

timogoebel commented:
@dbaars : Awesome, that this works. I do believe to know which bug you are still hitting, haven't had time to fix this yet. Could you confirm? It should be very easy.
Please do everything as you were doing before. When you see the "red" OS-Tab, please go to your JavaScript console and enter:

```javascript
bootdisk_provision_method_selected();
```

This should show you the validation error.
Another approach could be to simply select "Network Based" as a provision method and then select "Bootdisk Based" again.
Let me know of any progress :-)

dbaars commented:
Morning @timogoebel - this is what I get (this is on the built-from-scratch 3.0RC system, the one that bootdisk is not working on)

Run through "new host" setup, hit "Submit", Foreman/Katello comes back with the "red" OS-tab. In the javascript console I ran bootdisk_provision_method_selected();

_bootdisk_provision_method_selected();
undefined_

and in the GUI I see this

![image](https://cloud.githubusercontent.com/assets/14931710/14751976/dbde4bde-0920-11e6-88df-4d24eb568d1b.png)

timogoebel commented:
@dbaars : I just had a brief look into Katello's code and I believe, I spotted the problem.
Both foreman_bootdisk and Katello overwrite ```validate_media?``` in the Host Model and I guess, the wrong order is applied. I'm unsure however, how to fix this properly as I don't have a Katello install or dev setup.

What you can try as a workaround is the following:
Remove or comment
```ruby
alias_method_chain :validate_media?, :bootdisk
```
from ```app/models/concerns/foreman_bootdisk/host_ext.rb```

See:
https://github.com/Katello/katello/blob/b92a833e38fa3b39deff26839b219c0b7da53bbb/app/models/katello/concerns/host_managed_extensions.rb#L34-L36
https://github.com/theforeman/foreman_bootdisk/blob/master/app/models/concerns/foreman_bootdisk/host_ext.rb#L36-L38

dbaars commented:
Good morning @timogoebel :-)

That did the trick - I removed "alias_method_chain :validate_media?, :bootdisk" as suggested, rebooted, and tried creating a VM again - it worked!

I've noticed that sometimes the created VM will have the boot CD mounted, but seems to skip it and attempts boot off the network (which it can't do) - if I ctrl-alt-insert to reboot it, it will boot off the CD and carry on imaging, have you seen that behaviour?

Thanks,
Dylan

timogoebel commented:
@dbaars : Sorry, I have not seen that behaviour. I guess, that's a VMWare limitation as you cannot attach the cd as "connected at startup" but have to start the vm and then attach the cd via api.

I have created two Redmines for your issues:
http://projects.theforeman.org/issues/14797
http://projects.theforeman.org/issues/14796

dbaars commented:
Thanks @timogoebel. One (last?) question: is it possible for katello to use the shortname when creating the VM in vmare? The name in vmware is hostname.domain (e.g. test.squirrel.local) - can it be set to just create as "test" ?

Thanks!

dbaars commented:
Hi @timogoebel, opps another question for you! If this is the wrong place for these please tell me. When using "hammer host create", I can't see an option under --provision-method for bootdisk-based - if I try and create a host I get this error (I took a guess at "bootdisk" for the provision method)

```
Could not create the host:
  Error: option '--provision-method': value must be one of 'build', 'image'

  See: 'hammer host create --help'
```

I'd guess that "build" and "image" are the network based/image based methods?

Thanks :-)

timogoebel commented:
@dbaars:

> is it possible for katello to use the shortname when creating the VM in vmare?

Yes, the parameter is called "Use short name for VMs" in the "Provisioning" Tab in Settings.


> When using "hammer host create", I can't see an option under --provision-method for bootdisk-based

I'll take a look when I find the time.

> I'd guess that "build" and "image" are the network based/image based methods?

That's correct.

domcleal commented:
> If this is the wrong place for these please tell me

You'd be best using the users IRC channel or mailing list for questions, and if it's a bug then file it against either the Foreman or Bootdisk projects in Redmine as appropriate (http://projects.theforeman.org/projects/).
